### ISSUE TITLE
MLA decode is not batch-invariant on gfx950, capping greedy reproducibility and speculative target-equivalence

### ISSUE THREAD

**Issue by jhinpan** [lightseekorg/tokenspeed#904]
### Summary

The gfx950 MLA decode kernel is not batch-invariant: a row's output depends on how many other rows share the launch. This caps greedy reproducibility for continuous batching, and it caps greedy target-equivalence for **any** speculative algorithm on this backend, because target verify presents `bs * spec_num_tokens` rows where a plain decode presents `bs`.

### Root cause

The flash-decoding split count is chosen from the row count, not from per-row work:

https://github.com/lightseekorg/tokenspeed/blob/main/tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/attention/mla/decode.py

```python
_WAVE_WORKGROUPS = 256

def _select_num_kv_splits_bh16bn64(*, batch: int, max_seqlen_k: int, block_n: int) -> int:
    occupancy_cap = _WAVE_WORKGROUPS // batch
    blocks = (max_seqlen_k + block_n - 1) // block_n
    return max(1, min(occupancy_cap, blocks))
```

`batch` is `q.shape[0]`. Once `occupancy_cap` drops below `blocks`, the KV sequence is partitioned differently, and the fp32 online-softmax rescale in the split-combine kernel runs in a different order. Floating-point addition is not associative, so the logits differ in the low-order bits.

### Reproducer

Identical KV, identical page table, identical query in every row -- the launch shape is the only variable:

```
 seqlen  pages   max|delta| vs batch=1
    256      4   bs=2: 0.000e+00  bs=4: 0.000e+00  bs=8: 0.000e+00
   1024     16   bs=2: 0.000e+00  bs=4: 0.000e+00  bs=8: 0.000e+00
   2048     32   bs=2: 0.000e+00  bs=4: 0.000e+00  bs=8: 0.000e+00
   4096     64   bs=2: 0.000e+00  bs=4: 0.000e+00  bs=8: 9.766e-04
```

The boundary matches the formula exactly. At 64 pages the cap is 128 for `bs=1/2`, 64 for `bs=4` (both >= 64 blocks, so exact) and 32 for `bs=8` (below, so the reduction regroups).

Script: `test/runtime/kernels/test_mla_batch_invariance_probe.py` in #XXX asserts the exact cases and pins the boundary, so a kernel change that moves it is visible rather than surprising.

### Observed impact

Bringing up DSpark for Kimi-K3, greedy output matched a speculator-free run for the first 4 tokens across 24 prompts, then diverged (23/24 identical at 8 tokens, 20/24 at 16, 13/24 at 48).

Two controls show this is not the speculator:

- Committing only the target's own token every step (keeping the draft forward and the 8-token verify) reproduces the drift essentially unchanged: 19/24 vs 20/24 at 16 tokens, 12/24 vs 13/24 at 48.
- A 1-token draft drifts about as much as a 7-token one, which only makes sense if the trigger is `q_len != 1` rather than the number of accepted tokens.

Separately, at concurrency 8 a speculator-free server is itself only 5/8 reproducible run to run, because continuous batching varies each step's batch composition over the same non-invariant kernels.

Note the divergence is not confined to attention: an `lm_head`-shaped bf16 GEMM (163840 x 7168) already differs at `M=2` vs `M=1` on the same input row, and it is the projection that decides the greedy token. So attention is one contributor rather than the whole story.

### Why it matters

- Any speculative algorithm (EAGLE3, MTP, DFLASH, DSPARK) inherits this, so "verify speculative output is target-equivalent under deterministic greedy decoding" is not achievable as a bitwise property on this backend. Correctness gates should use distributional/accuracy equivalence instead, which is what the literature does ([MLSys 2026](https://arxiv.org/pdf/2601.11580) reports the same for vLLM and attributes it to numerical nondeterminism rather than implementation bugs).
- Continuous batching is not greedy-reproducible at concurrency > 1, which matters for eval harnesses, LLM-as-judge, and RL rollouts.

### Possible directions

1. Document it, and state the equivalence guarantee as distributional. Zero cost, and matches [NeurIPS 2025](https://proceedings.neurips.cc/paper_files/paper/2025/file/f80094a824ba5912d4a2de169c404a40-Paper-Conference.pdf) / batch-invariant-kernels prior art.
2. Offer an opt-in deterministic mode that fixes `num_kv_splits` independently of `batch` (and correspondingly fixes GEMM split-K), accepting the occupancy loss. Useful for RL and evaluation, off by default.
3. Full batch-invariant kernels, as in the prior art above. Correct but costly, and a larger piece of work than this issue.

Happy to prototype (2) behind a flag if that is the preferred direction.
